File: fex_0007_amba_axi_xfft/iplib/xfft_config/api/c/xfft_config_api.c
//------------------------------------------------------------------------------
// Copyright (c) 2019 by Future Design Systems
// All right reserved.
//
// http://www.future-ds.com
//------------------------------------------------------------------------------
// VERSION = 2019.04.05.
//------------------------------------------------------------------------------
#include "conapi.h"
#include "xfft_config_api.h"
#include "memory_map.h"
extern con_Handle_t handle;

//------------------------------------------------------------------------------
// Register access macros
#if defined(TRX_BFM)||defined(TRX_AXI)||defined(TRX_AHB)
#include <stdio.h>
#include "bfm_api.h"
#   define REGRD(A,B)         BfmRead(handle, (unsigned int)(A), (unsigned int*)&(B), 4, 1);
#   define REGWR(A,B)         BfmWrite(handle, (unsigned int)(A), (unsigned int*)&(B), 4, 1);
#	define   uart_put_string(x)  printf("%s", (x));
#	define   uart_put_hex(n)  printf("%x", (n));
#else
#include "uart_api.h"
#       define REGWR(A, B)   *(unsigned *)A = B;
#       define REGRD(A, B)    B = *(unsigned *)A;
#endif

//------------------------------------------------------------------------------
#ifndef ADDR_XFFT_CONFIG_START
#error  ADDR_XFFT_CONFIG_START should be defined
#endif

#define XFFT_CONFIG_VERSION (ADDR_XFFT_CONFIG_START+0x00)
#define XFFT_CONFIG_RESET   (ADDR_XFFT_CONFIG_START+0x10)
#define XFFT_CONFIG_CONFIG  (ADDR_XFFT_CONFIG_START+0x14)
#define XFFT_CONFIG_STATUS  (ADDR_XFFT_CONFIG_START+0x18)

//------------------------------------------------------------------------------
// writing 1 to bit 0  causes XFFT_ARESETn 0.
// writing 0 to bit 0  causes XFFT_ARESETn 1.
int xfft_config_reset(void)
{
     volatile uint32_t value = 1;
     REGWR(XFFT_CONFIG_RESET,value);
     value = 0;
     REGWR(XFFT_CONFIG_RESET,value);
     return 0;
}

//------------------------------------------------------------------------------
// status[3] = done
// status[2] = tvalid
// status[1] = tready
// status[0] = resetn
int xfft_config( unsigned int config, int time_out )
{
     volatile uint32_t value;
     REGWR(XFFT_CONFIG_CONFIG, config);
     int num=0;
     do { REGRD(XFFT_CONFIG_STATUS,value);
          if ((value&0x9)==0x9) break; // done & ~reset
          num++;
     } while ((time_out==0)||(num<time_out));
     if ((time_out!=0)&&(num>=time_out)) return -1;
     return 0;
}

//--------------------------------------------------------
#undef REGWR
#undef REGRD

//------------------------------------------------------------------------------
// Revision History
//
// 2019.04.05: started by Ando Ki (adki@future-ds.com)
//------------------------------------------------------------------------------

File: fex_0007_amba_axi_xfft/iplib/xfft_config/api/python/xfft_config_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This file contains Python version of XFFT_CONFIG with CON-FMC AMBA BFM.
"""
__author__     = "Ando Ki"
__copyright__  = "Copyright 2019, Future Design Systems"
__credits__    = ["none", "some"]
__license__    = "FUTURE DESIGN SYSTEMS SOFTWARE END-USER LICENSE AGREEMENT FOR CON-FMC."
__version__    = "1"
__revision__   = "0"
__maintainer__ = "Ando Ki"
__email__      = "contact@future-ds.com"
__status__     = "Development"
__date__       = "2019.04.15"
__description__= "XFFT_CONFIG with CON-FMC AMBA BFM"

#-------------------------------------------------------------------------------
class xfft_config:
      def __init__(self, hdl, bfm, addr=0x00000000):
          self._hdl   = hdl
          self._bfm   = bfm
          self._VERSION=(addr+0x00)
          self._RESET  =(addr+0x10)
          self._CONFIG =(addr+0x14)
          self._STATUS =(addr+0x18)

      def csr( self ):
          value = [0]
          self._bfm.BfmRead(self._hdl, self._VERSION, value); print "_VERSION: ", hex(value[0])
          self._bfm.BfmRead(self._hdl, self._RESET  , value); print "_RESET  : ", hex(value[0])
          self._bfm.BfmRead(self._hdl, self._CONFIG , value); print "_CONFIG : ", hex(value[0])
          self._bfm.BfmRead(self._hdl, self._STATUS , value); print "_STATUS : ", hex(value[0])

      def reset( self ):
          """writing 1 to bit 0  causes XFFT_ARESETn 0.
             writing 0 to bit 0  causes XFFT_ARESETn 1."""
          self._bfm.BfmWrite(self._hdl, self._RESET,[0x1])
          self._bfm.BfmWrite(self._hdl, self._RESET,[0x0])
          return 0

      def config( self, config, time_out ):
          """status[3] = done
             status[2] = tvalid
             status[1] = tready
             status[0] = resetn"""
          value = [config]
          self._bfm.BfmWrite(self._hdl, self._CONFIG, value)
          num=0
          value = [0]
          while True:
                self._bfm.BfmRead(self._hdl, self._STATUS, value);
                if ((value[0]&0x9)==0x9): break # done & ~reset
                num += 1
                if ((time_out!=0) and (num>=time_out)): break
          if (time_out!=0) and (num>=time_out): return -1;
          return 0;

#===============================================================================
# Revision history:
#
# 2019.04.15: Started by Ando Ki (adki@future-ds.com)
#             - Not finished yet
#===============================================================================
